Patch hdd images before migration

dgutu-suse opened the pull request:
Based on - https://progress.opensuse.org/issues/10996
Couple of examples:
http://corr.arch.suse.de/tests/1048
http://corr.arch.suse.de/tests/944
http://wingenfelder.arch.suse.de/tests/44
This PR will also require changes on testcases related to offline migrations.

okurz commented on lib/bootloader_setup.pm:
```diff
@@ -65,6 +65,7 @@ sub select_bootmenu_option {
     }
     if (get_var('ZDUP') || get_var("ONLINE_MIGRATION")) {
         send_key 'ret';    # boot from hard disk
+        assert_screen 'grub2', 15;
```
why?

dgutu-suse replied:
Perhaps I wanted to catch the grub menu :) It safe to remove.

okurz commented on lib/registration.pm:
```diff
@@ -201,6 +201,7 @@ sub fill_in_registration_data {
             assert_screen("module-selection");
             send_key $cmd{next};
         }
+        else {send_key $cmd{next};}
```
that does not make any sense. If the same key is pressed in both "if" as well as "else", it should be outside the if-else.
OTOH why? ;-)

dgutu-suse replied:
I've found why I was using this next. Its safe to be removed.

okurz commented on lib/utils.pm:
```diff
@@ -213,7 +213,7 @@ sub fully_patch_system {
     my (%args) = @_;
     $args{version_variable} //= 'VERSION';
-    if (sle_version_at_least('12-SP1', version_variable => $args{version_variable})) {
+    if (sle_version_at_least('12-SP1', version_variable => $args{version_variable}) && !check_var('HDDVERSION', 'SLES-11')) {
```
I don't think this is a good idea. It already tests for sle_version_at_least with a forward of the type of variable so it should do exactly what you want to do already

dgutu-suse replied:
Removing, changed mainly to handle sles11

okurz commented on lib/utils.pm:
```diff
@@ -331,19 +331,26 @@ sub ensure_shim_import {
         assert_and_click 'reboot-auth-showtext';         # Click the 'Show Text' Option to enable the display of the typed text
         assert_screen 'reboot-auth-correct-password';    # Check the password is correct
-
+    }
```
that is messed up style

dgutu-suse replied:
Removing, changed mainly to handle sles11

okurz commented on products/sle/main.pm:
```diff
@@ -969,7 +969,12 @@ sub load_fips_tests_crypt() {
+    loadtest "update/patch_before_migration";
+    loadtest "console/consoletest_finish_sym";
+    loadtest "x11/reboot_and_install";
+}
```
please use only 'single ticks' with loadtest unless you want to use variable expansion

dgutu-suse replied:
ok

okurz commented on tests/autoyast/autoyast_reboot.pm:
```diff
@@ -25,8 +25,12 @@ sub run {
+    if (check_var("BOOTFROM", "d")) {
+        assert_screen("inst-bootmenu", 60);
+    }
+    else { assert_screen("autoyast-boot", 20); }
```
the change in this file seems unrelated. Please move into another commit.

dgutu-suse replied:
Actually this change is very related because autoupgrade jobs are also migration and I think they also deserves to be patched :) If autoupgrade are not migrations then I will revert changes to this file.

okurz commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,108 @@
+use registration;
+
+sub patching_sle11() {
+
```
delete blank line

dgutu-suse replied:
ok

okurz commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,108 @@
+    save_screenshot;
+    assert_script_run('zypper lr -d');
+
+    #Patch the system
```
redundant comment, the next function call has a good name already :-)

dgutu-suse replied:
+1

okurz commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,108 @@
+    my $reg_out = script_output("suse_register -E");
+    die "Unable to erase system registration data" unless $reg_out =~ /Successfully erased local registration data/;
+    save_screenshot;
+    script_run("gconftool-2 --set /apps/gnome-packagekit/frequency_get_updates --type=string daily");
```
why?

dgutu-suse replied:
Removing, was for sle11 to disable refreshing of repos after registering.

okurz commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,108 @@
+}
+
+sub patching_sle12() {
+
```
delete blank line

dgutu-suse replied:
ok

okurz commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,108 @@
+    assert_script_run('zypper lr -d');
+}
+
+sub patching_sle12() {
```
isn't that like a generic patch function, i.e. `patching_sle`?

dgutu-suse replied:
+1

okurz commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,108 @@
+}
+
+sub run() {
+
```
delete blank line

dgutu-suse replied:
ok

okurz commented on tests/x11/reboot_and_install.pm:
```diff
@@ -15,14 +15,16 @@ use strict;
 sub run() {
     reboot_gnome;    # reboot from previously booted hdd to do pre check or change e.g. before upgrade
-    select_bootmenu_option;
+    return if select_bootmenu_option == 3;
```
what does '3' mean?

dgutu-suse replied:
it is evaluation of  lib/bootloader_setup   
###
if (get_var('ZDUP') || get_var("ONLINE_MIGRATION")) {                                                                                     send_key 'ret';    # boot from hard disk                                                                                              
return 3;
###

okurz commented on tests/x11/reboot_and_install.pm:
```diff
@@ -15,14 +15,16 @@ use strict;
-    # boot
-    send_key "ret";
+    specific_bootmenu_params;
+    registration_bootloader_params(utils::VERY_SLOW_TYPING_SPEED);
```
why?

dgutu-suse replied:
Because of my (bad) qemu  setup, its safe to remove.

okurz commented on tests/x11/reboot_and_install.pm:
```diff
@@ -15,14 +15,16 @@ use strict;
-    send_key "ret";
+    specific_bootmenu_params;
+    registration_bootloader_params(utils::VERY_SLOW_TYPING_SPEED);
+    send_key "ret";    #boot
```
better keep it atop the current line as before

dgutu-suse replied:
ok

dgutu-suse commented:
Thanks for comment, I'll took a look what needs to be changed.
Yes, the changes are not small, and potentially something can fail but since I've modified files which has effect on other testcases like yast and online migrations, I have checked their run.

okurz commented:
Any luck on updating? Just smaller changes for a start would be nice :-)

okurz commented:
@dgutu-suse please update your PR with commits that focus on one aspect first so that we can merge them soon. Then we can continue with the next steps afterwards. I would like to see your changes in "master" so that we can make real use of it and others help maintaining it.

mitiao commented on tests/boot/boot_to_desktop_sym.pm:
```diff
@@ -0,0 +1 @@
+boot_to_desktop.pm
```
why not directly call boot_to_desktop?

dgutu-suse replied:
When a test is called 2 times the last is skipped.

mitiao commented on tests/console/consoletest_finish_sym.pm:
```diff
@@ -0,0 +1 @@
+consoletest_finish.pm
```
call consoletest_finish directly

dgutu-suse replied:
When a test is called 2 times the last is skipped.

mitiao commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,68 @@
+    save_screenshot;
+    yast_scc_registration();
+    assert_script_run('zypper lr -d');
+    minimal_patch_system(version_variable => 'HDDVERSION');
```
this only apply minimal patch, i don't see fully patch scenario

dgutu-suse replied:
Thank you for taking a look :) I'm adding the variable to control the minimal and full patching.

mitiao commented:
Seems you have implemented upgrading on patched system by autoyast and zdup, what about upgrading by DVD?

dgutu-suse commented:
@mitiao, can you please rephrase? autoupgrade by yast, zdup are all upgrade from DVD or do you mean another thing?

mitiao commented:
testing like this: 
https://openqa.suse.de/tests/773488
but patched before upgrading.

dgutu-suse commented:
@mitiao - yes, the one you show is not patching anything, here is the same example on my openqa instance: http://corr.arch.suse.de/tests/1514

okurz commented on lib/utils.pm:
```diff
@@ -343,9 +343,8 @@ sub reboot_gnome {
-        # we need to kill ssh for iucvconn here,
-        # because after pressing return, the system is down
+                                                         # we need to kill ssh for iucvconn here,
+                                                         # because after pressing return, the system is down
```
this comment indendation looks weird, please re-add the blank line before

okurz commented on products/sle/main.pm:
```diff
@@ -969,7 +969,12 @@ sub load_fips_tests_crypt() {
     loadtest "console/ecryptfs_fips";
 }
-
+sub load_patching_tests() {
```
please add blank line before

okurz commented on products/sle/main.pm:
```diff
@@ -969,7 +969,12 @@ sub load_fips_tests_crypt() {
+    loadtest 'console/consoletest_finish_sym';
+    loadtest 'x11/reboot_and_install';
+}
 sub prepare_target() {
```
please add blank line in between

okurz commented on tests/console/consoletest_finish.pm:
```diff
@@ -24,9 +24,7 @@ sub run() {
     save_screenshot();
-
     script_run "systemctl unmask packagekit.service";
-
```
please keep blank lines as before

mitiao commented:
with your given instance: http://corr.arch.suse.de/tests/1514, i think it is patched offline migration by zdup.
the scenario by DVD I expect is:
patch system -> reboot -> select 'Upgrade' in bootmenu -> perform upgrading by yast installer

dgutu-suse commented:
@mitiao, my bad, I gave you the zdup case. Here is the migration from installer - http://corr.arch.suse.de/tests/1506

QingmingSu commented:
Please don't make it a mandatory action to patch (update) the image before migrating. Because sometimes we'd like to test migration from an image with no patch (update).

dgutu-suse commented:
@QingmingSu, the original hdd image will remain intact so you are able to re-use it for your purposes. Does my explanation make send to you?

dgutu-suse commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,76 @@
+
+    set_var("VIDEOMODE",    'text');
+    set_var("SCC_REGISTER", 'installation');
+    set_var("SP2ORLATER",   '0') if (!sle_version_at_least('12-SP2', version_variable => 'HDDVERSION'));
```
@mitiao, I've encounter a problem to register scc addons in text mode after your changes here - https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/2138 
Without this I have troubles for sles12ga and sles12sp1.

mitiao replied:
version_variable => 'HDDVERSION' indicates that we are going to do registration on HDD images, so set_var should be HDD_SP2ORLATER.

dgutu-suse replied:
It's even save to be removed. Thx for noticing it.

dgutu-suse commented on tests/installation/zdup.pm:
```diff
@@ -90,6 +90,7 @@ sub run() {
                        [ -z \$dev ] || echo \"found dev \$dev with label \$label\"";
             # if that fails, e.g. if volume descriptor too long, just try /dev/sr0
             $defaultrepo = "dvd:/?devices=\${dev:-/dev/sr0}";
+            $defaultrepo = "cd:/?devices=\${dev:-/dev/sr0}" if get_var('PATCH');
```
Without this change zypper_lr is complaining about missing repo and test fail. Could this be related to the fact that this testcase will use BOOTFROM=d variable?

dgutu-suse replied:
I'm gonna remove this "if" because the issue is in my openqa environment.

sysrich commented:
@QingmingSu the SLE PRD's make it quite clear that the expected supported behaviour is to support migrations with either a minimal patch (just the update stack) or a full patch.

I can find no mention in the PRD of supporting migrations without patching - if we don't support it, why do we need to test it?

sysrich commented:
@dgutu-suse is this still WIP/notready?

okurz commented on lib/utils.pm:
```diff
@@ -395,6 +395,7 @@ sub reboot_gnome {
     assert_screen 'logoutdialog';
     assert_and_click 'logoutdialog-reboot-highlighted';
 
+
```
unrelated change

dgutu-suse replied:
Removing...

okurz commented on tests/console/consoletest_finish.pm:
```diff
@@ -24,7 +24,6 @@ sub run() {
     type_string "loginctl --no-pager\n";
     wait_still_screen(2);
     save_screenshot();
-
```
unrelated change

dgutu-suse replied:
Adding back then.

okurz commented on tests/installation/zdup.pm:
```diff
@@ -65,7 +65,7 @@ sub run() {
     else {
         #SUSEMIRROR not set, zdup from ftp source for online migration
-        if (check_var('TEST', "migration_zdup_online_sle12_ga")) {
+        if (get_var('TEST') =~ /migration_zdup_online_sle12_ga/) {
```
why not keep the old?

dgutu-suse replied:
Because every migration testcase will have fullupdate and minimalupdate in their names, or shorten one like mu (minimalupdate) and fu (fullupdate)

okurz commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,76 @@
+use registration;
+
+sub patching_sle() {
+
```
delete blank line

dgutu-suse replied:
OK.

okurz commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,76 @@
+}
+
+sub run() {
+
```
delete blank line

dgutu-suse replied:
Am I fan of blank lines ? :) Removing.

okurz commented on tests/update/patch_before_migration.pm:
```diff
@@ -0,0 +1,76 @@
+
+    select_console 'root-console';
+
+    type_string "chown $username /dev/$serialdev\n";
```
this is repeating what is also necessary in consoletest_setup. Please extract method

dgutu-suse replied:
ok

QingmingSu commented:
@sysrich, Agree that it makes more sense to always update the system before migrating to newer version.

dgutu-suse commented:
@okurz @sysrich @mitiao @QingmingSu thx for reviewing. I have only one things that stops me to remove the labels. On my openqa instance the mounted installation source (like sle12sp3 dvd) is always a CD-ROM but as soon as I see on o.s.d is always DVD-ROM so my runs failing on zypper_lr because of mismatching or repos. So I've removed this part - $defaultrepo = "cd:/?devices=\${dev:-/dev/sr0}" if get_var('PATCH'); I get also DVD-ROM when booting the same hdd image through virt-manager so there is something that makes it CD-ROM in my openqa instance. 
Some examples:
http://corr.arch.suse.de/tests/1746#step/zypper_lr/13 - here the source was still mounted as CD-ROM that's why it passing the zdup test but since I removed the if above it will start to fail on zdup testcase.

dgutu-suse commented:
All above applying only to 'zdup'. autoupgrade and simple offline migration is installed from CD-ROM.

sysrich commented:
@dgutu-suse what is the question? - your requestion from 13 days ago doesn't have a question mark in it (besides the the defaultrepo string) - so I can't identify what you want answered.

sysrich commented:
on hold till @dgutu-suse is back from FTO, has a modification he wishes to make

dgutu-suse commented:
It's ok to be merged.
